trace_replay/trace_replay.cc

```diff
@@ -126,222 +126,218 @@ bool TracerHelper::SetPayloadMap(uint64_t& payload_map,
-  } else {
-    Slice buf(trace->payload);
-    GetFixed64(&buf, &trace->payload_map);
-    int64_t payload_map = static_cast<int64_t>(trace->payload_map);
-    while (payload_map) {
-      // Find the rightmost set bit.
-      uint32_t set_pos =
-          static_cast<uint32_t>(log2(payload_map & -payload_map));
-      switch (set_pos) {
-        case TracePayloadType::kGetCFID:
-          GetFixed32(&buf, &cf_id);
-          break;
-        case TracePayloadType::kGetKey:
-          GetLengthPrefixedSlice(&buf, &get_key);
-          break;
-        default:
-          assert(false);
+      if (record != nullptr) {
+        record->reset(new WriteQueryTraceRecord(std::move(rep), trace->ts));
       }
-      // unset the rightmost bit.
-      payload_map &= (payload_map - 1);
-    }
-  }
 
-  if (record != nullptr) {
-    PinnableSlice ps;
-    ps.PinSelf(get_key);
-    record->reset(new GetQueryTraceRecord(cf_id, std::move(ps), trace->ts));
-  }
-
-  return Status::OK();
-}
+      return Status::OK();
+    }
+    // Get
+    case kTraceGet: {
+      uint32_t cf_id = 0;
+      Slice get_key;
+
+      if (trace_file_version < 2) {
+        DecodeCFAndKey(trace->payload, &cf_id, &get_key);
+      } else {
+        Slice buf(trace->payload);
+        GetFixed64(&buf, &trace->payload_map);
+        int64_t payload_map = static_cast<int64_t>(trace->payload_map);
+        while (payload_map) {
+          // Find the rightmost set bit.
+          uint32_t set_pos =
+              static_cast<uint32_t>(log2(payload_map & -payload_map));
+          switch (set_pos) {
+            case TracePayloadType::kGetCFID: {
+              GetFixed32(&buf, &cf_id);
+              break;
+            }
+            case TracePayloadType::kGetKey: {
+              GetLengthPrefixedSlice(&buf, &get_key);
+              break;
+            }
+            default: {
+              assert(false);
+            }
+          }
+          // unset the rightmost bit.
+          payload_map &= (payload_map - 1);
+        }
+      }
 
-Status TracerHelper::DecodeIterRecord(Trace* trace, int trace_file_version,
-                                      std::unique_ptr<TraceRecord>* record) {
-  assert(trace != nullptr);
-  assert(trace->type == kTraceIteratorSeek ||
-         trace->type == kTraceIteratorSeekForPrev);
+      if (record != nullptr) {
+        PinnableSlice ps;
+        ps.PinSelf(get_key);
+        record->reset(new GetQueryTraceRecord(cf_id, std::move(ps), trace->ts));
+      }
 
-  if (record != nullptr) {
-    record->reset(nullptr);
-  }
+      return Status::OK();
+    }
+    // Iterator Seek and SeekForPrev
+    case kTraceIteratorSeek:
+    case kTraceIteratorSeekForPrev: {
+      uint32_t cf_id = 0;
+      Slice iter_key;
+      Slice lower_bound;
+      Slice upper_bound;
+
+      if (trace_file_version < 2) {
+        DecodeCFAndKey(trace->payload, &cf_id, &iter_key);
+      } else {
+        Slice buf(trace->payload);
```

Comment:
It is not that elegant to put the payload decoding in the trace analyzer.......

Reply:
This is in TracerHelper, not in TraceAnalyzer. I merged DecodeWriteTraceRecord / DecodeGetTraceRecord, ... into a single `DecodeTraceRecord`, then you don't need to call each individual decoding functions in `TraceAnalyzer` and `ReplayerImpl`.

Reply:
Sorry, I did not notice that....